### PR TITLE
Data Product Catalog prototype

### PR DESCRIPTION
**Pull request by timcanham**
| | |
|:---|:---|
|**_Related Issue(s)_**| https://github.com/nasa/fprime/issues/1713 |
|**_Has Unit Tests (y/n)_**| y |
|**_Documentation Included (y/n)_**| y |

---
## Change Description

This is a prototype demonstration of `Svc/DpCatalog` It was used to demonstrate an early release of data products in conjunction with [FPP updates](https://github.com/fprime-community/fpp/issues?q=is%3Aissue+is%3Aopen++%5BData+Products%5D) and [`DpManager/DpWriter`](https://github.com/nasa/fprime/pulls?q=is%3Apr+is%3Aclosed+data+product)

`Svc/DpCatalog` is not complete yet - this version is to capture the demo and allow users to play with it.

## Rationale

Get a consolidated set of data product features int he baseline.

## Testing/Review Recommendations

The demo can be run and experimented with.

## Future Work

`Svc/DpCatalog` work will be continued with further releases.

### CODE REVIEW

**Comment by timcanham** [comment]
Requested @bocchino for changes made to `Svc/DpWriter`.

File: Svc/DpCatalog/DpCatalog.cpp
Comment (Joshua-Anderson) [inline Svc/DpCatalog/DpCatalog.cpp:161]: Might be worth lowering the severity of this EVR to diagnostic, sending a event per data product file processed could add up.

**Comment by Joshua-Anderson** [comment]
Is storing transmit state (UNTRANSMITTED/PARTIAL/TRANSMITTED) in the data product header the long term term plan?

It seems like this would force us to re-write data products once we start transmitting them, which could be a bit of a circular dependency. It also could get a bit confusing because the onboard DP file checksum would change after you start transmitting it.

Did we consider having DpCatalog own data product transmit state instead? (I.E. have a catalog file with a list of data products and their transmit states)

This could also provide a bit more flexibility because different projects may want to track data product transmit state differently (I.E. track not just that a data product is partial but how much of the file was sent, or which portions of the file need to be retransmitted).

**Comment by timcanham** [comment]
@Joshua-Anderson The plan is to modify the header as we process the data products, i.e. `DpWriter` writes it as `UNTRANSMITTED`, then as `DpCatalog` starts downlinking it will change to `PARTIAL`, then `TRANSMITTED` when done. I'm not sure I see the circular dependency part, so elaborate on that for me.


You are correct that having a separate DP state file (or other variants on the idea) is an option and @bocchino and I discussed it, but I've always preferred files to be self-contained in their state so that the `DpCatalog` can be simpler in that it discovers state in the files and doesn't have to keep track of things in two places. Just a design preference.

File: config/DpCfg.hpp
Comment (bocchino) [inline config/DpCfg.hpp]: Let's add the new string field to the comment in line 19.

File: Svc/DpWriter/DpWriter.hpp
Comment (bocchino) [inline Svc/DpWriter/DpWriter.hpp]: Maybe this should be Fw::FileNameString?

**Comment by bocchino** [comment]
The changes to DpWriter look good to me! I just had a couple of comments.

File: config/DpCfg.hpp
Comment (timcanham) [reply config/DpCfg.hpp]: Added comment.

File: Svc/DpWriter/DpWriter.hpp
Comment (timcanham) [reply Svc/DpWriter/DpWriter.hpp]: Changed them to `Fw::FileNameString`

Comment (bocchino) [reply Svc/DpWriter/DpWriter.hpp]: Looks good. In my edits I changed the type in the configure prototype to `const Fw::StringBase&`. I think we should use this pattern going forward whenever the concrete string type is not needed.

File: config/DpCfg.hpp
Comment (bocchino) [reply config/DpCfg.hpp]: Looks good!

File: Ref/SignalGen/SignalGen.fpp
Comment (LeStarch) [inline Ref/SignalGen/SignalGen.fpp:42]: Why is there bothj get and request on this component and one is left in TODO state?

File: Ref/SignalGen/SignalGen.cpp
Comment (LeStarch) [inline Ref/SignalGen/SignalGen.cpp:255]: This should be removed or demonstrated.

**Review by LeStarch — COMMENTED** [review]
@bocchino would you mark a final review on this demo component?  Just clarify your comments were resolved.

File: Ref/SignalGen/SignalGen.cpp
Comment (timcanham) [reply Ref/SignalGen/SignalGen.cpp:255]: I was going to add that update to `SignalGen` after the demo was merged.

**Review by bocchino — APPROVED** [review]
My comments have been addressed.

File: Ref/SignalGen/SignalGen.cpp
Comment (timcanham) [reply Ref/SignalGen/SignalGen.cpp:255]: Hold off for a day or so - decided to implement the missing function.

Comment (LeStarch) [reply Ref/SignalGen/SignalGen.cpp:255]: @timcanham this was merged yesterday.

Comment (timcanham) [reply Ref/SignalGen/SignalGen.cpp:255]: Ah... Okay, I'll do another PR for the update.

**Pull request merged**
